src/Aspire.Hosting/Publishing/Internal/DeploymentStateManagerBase.cs

```diff
@@ -0,0 +1,275 @@
+    /// <returns>The loaded state as a JsonObject.</returns>
+    protected async Task<JsonObject> LoadStateAsync(CancellationToken cancellationToken = default)
+    {
+        await _loadLock.WaitAsync(cancellationToken).ConfigureAwait(false);
+        try
+        {
+            if (_isStateLoaded && _state is not null)
+            {
+                return _state;
+            }
+
+            var jsonDocumentOptions = new JsonDocumentOptions
+            {
+                CommentHandling = JsonCommentHandling.Skip,
+                AllowTrailingCommas = true,
+            };
+
+            var statePath = GetStatePath();
+
+            if (statePath is not null && File.Exists(statePath))
+            {
+                var fileContent = await File.ReadAllTextAsync(statePath, cancellationToken).ConfigureAwait(false);
+                var flattenedState = JsonNode.Parse(fileContent, documentOptions: jsonDocumentOptions)!.AsObject();
+                _state = UnflattenJsonObject(flattenedState);
+            }
+            else
+            {
+                _state = [];
+            }
+
+            _isStateLoaded = true;
+            return _state;
+        }
+        finally
+        {
+            _loadLock.Release();
+        }
+    }
+
+    /// <inheritdoc/>
+    public async Task SaveStateAsync(JsonObject state, CancellationToken cancellationToken = default)
+    {
+        await _saveLock.WaitAsync(cancellationToken).ConfigureAwait(false);
+        try
+        {
+            await SaveStateToStorageAsync(state, cancellationToken).ConfigureAwait(false);
+        }
+        finally
+        {
+            _saveLock.Release();
+        }
+    }
+
+    private SectionMetadata GetSectionMetadata(string sectionName)
+    {
+        return _sections.GetOrAdd(sectionName, _ => new SectionMetadata());
+    }
+
+    /// <inheritdoc/>
+    public async Task<DeploymentStateSection> AcquireSectionAsync(string sectionName, CancellationToken cancellationToken = default)
+    {
+        await LoadStateAsync(cancellationToken).ConfigureAwait(false);
+
+        var metadata = GetSectionMetadata(sectionName);
+
+        var sectionData = _state?.TryGetPropertyValue(sectionName, out var sectionNode) == true && sectionNode is JsonObject obj
+            ? obj.DeepClone().AsObject()
+            : null;
+
+        return new DeploymentStateSection(sectionName, sectionData, metadata.Version);
+    }
+
+    /// <inheritdoc/>
+    public async Task SaveSectionAsync(DeploymentStateSection section, CancellationToken cancellationToken = default)
+    {
+        await LoadStateAsync(cancellationToken).ConfigureAwait(false);
+
+        if (_state is null)
+        {
+            throw new InvalidOperationException("State has not been loaded.");
+        }
+
+        // Atomically check version and update using AddOrUpdate (compare-exchange pattern)
+        var versionMismatch = false;
+        _sections.AddOrUpdate(
```

Comment:
IMO this should use a regular lock + dictionary. That would make the semantics much clearer. Eg, the `updateValueFactory` can be called concurrently, so this code here can concurrently mutate `existing`. See https://github.com/dotnet/runtime/blob/3dc26a3e705316577b432fb5286c9c4498d8ea22/src/libraries/System.Private.CoreLib/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L1325-L1354

Since the return value from the `updateValueFactory` is the existing value and the existing value is passed to `TryUpdateInternal` and used as the `comparisonValue`, I think this will not provide the intended guarantees.

Best to use a lock + Dictionary. For good measure, you could make SectionMetadata read-only.